csatar-plugins/knowledgerepository/classes/xlsx/SongsXlsxImport.php

```diff
@@ -0,0 +1,224 @@
+use Flash;
+use Lang;
+use Maatwebsite\Excel\Concerns\Importable;
+use Maatwebsite\Excel\Concerns\OnEachRow;
+use Maatwebsite\Excel\Concerns\RemembersRowNumber;
+use Maatwebsite\Excel\Concerns\SkipsEmptyRows;
+use Maatwebsite\Excel\Concerns\SkipsFailures;
+use Maatwebsite\Excel\Concerns\SkipsOnFailure;
+use Maatwebsite\Excel\Concerns\SkipsUnknownSheets;
+use Maatwebsite\Excel\Concerns\WithGroupedHeadingRow;
+use Maatwebsite\Excel\Concerns\WithHeadingRow;
+use Maatwebsite\Excel\Concerns\WithMultipleSheets;
+use Maatwebsite\Excel\Row;
+use PhpOffice\PhpSpreadsheet\IOFactory;
+use PhpOffice\PhpSpreadsheet\Spreadsheet;
+
+class SongsXlsxImport implements OnEachRow, WithHeadingRow, WithGroupedHeadingRow, SkipsOnFailure, WithMultipleSheets, SkipsUnknownSheets
+{
+    use Importable, RemembersRowNumber, SkipsFailures;
+
+    private $associationId;
+
+    private $uploaderCsatarCode;
+
+    private $approverCsatarCode;
+
+    private $overwrite = false;
+
+    private $effectiveKnowledgeOnly = false;
+
+    public $errors = [];
+
+    private $worksheetRaw;
+
+    public function __construct($associationId, $overwrite, $richTextColumns, $worksheetRaw, $uploaderCsatarCode, $approverCsatarCode)
+    {
+        $this->associationId = $associationId;
+        $this->overwrite = $overwrite;
+        $this->richTextColumns = $richTextColumns;
+        $this->worksheetRaw = $worksheetRaw;
+        $this->uploaderCsatarCode = $uploaderCsatarCode;
+        $this->approverCsatarCode = $approverCsatarCode ?? null;
+        $this->approved_at = $approverCsatarCode ? date('Y-m-d H:i:s') : null;
+    }
+
+    public function sheets(): array
+    {
+        return [
+            0 => $this,
+        ];
+    }
+
+    public function onUnknownSheet($sheetName)
+    {
+        // E.g. you can log that a sheet was not found.
+        info("Sheet {$sheetName} was skipped");
+    }
+
+    public function onRow(Row $row)
+    {
+        $cellsArray = $row->toArray();
+
+        if (!isset($cellsArray['sorszam']) || !isset($cellsArray['dal_cim'])) {
+            return null;
+        }
+
+        $song = Song::where('association_id', $this->associationId)
+            ->where('title', str_replace("\n", "", $cellsArray['dal_cim']))
+            ->first();
+
+        if (!$this->overwrite && !empty($song)) {
+            $this->errors[$row->getRowIndex()][] = Lang::get('csatar.knowledgerepository::lang.plugin.admin.song.songAlreadyExists', ['title' => $cellsArray['dal_cim']]);
+            return;
+        }
+
+        $this->convertRichTextToHtml($cellsArray, $row, $this->richTextColumns);
+
+        if (empty($song)) {
+            $song = new Song();
+            $song->association_id = $this->associationId;
+            $song->title = $cellsArray['dal_cim'];
+        }
+
+        if (isset($cellsArray['dal_tipusa'])) {
+            $song->song_type_id = $this->getModelIds($row, $cellsArray['dal_tipusa'], SongType::class, 'name', null, null, true)[0] ?? null;
+        }
```

Comment:
Esetleg az if-et is be lehetne tenni a getModelId fuggvenybe, hogy ne kelledjen itt tobbszor meghivni, nem?

Reply:
Betettem, bár nem tudom, hogy melyik kerül többe... Az if és függvény hívás, ha átmegy az if, vagy a függvényhívás és az if és megnézni, hogy `$cellsArray['dal_tipusa']` létezik-e.